roachtest: create TPC-C benchmarking tool

nvb opened the pull request:
Closes #25084.

This change creates a new `tpccbench` roachtest specifically for
the purpose of benchmarking TPC-C. The test spins up a cluster (if
not provided a `--cluster` flag), restores a TPC-C dataset (if not
run on an existing cluster with `--wipe=false` specified), and
benchmarks the cluster's ability to scale under an increasing number
of TPC-C warehouses. Specifically, it uses a line search algorithm
to efficiently search for the maximum number of warehouses that can
be operated on while sustaining a throughput threshold, which is set
to a fraction of max tpmC for that warehouse count.

@a-robinson I pushed this WIP change so that you and others can
play around with it and give feedback on its approach and ergonomics.

cc. @jordanlewis I'd love to hear your thoughts on the early draft of this
tool and if you have any suggestions for how it can be improved or made
more useful.

Release note: None

nvb committed:
util/search: create new util/search package

This package includes a Searcher interface, which is implemented by
a binarySearcher type and a lineSearcher type. The interface
can be used to search for the largest value that passes a given
predicate function. The two implementations provide different
algorithms for achieving this goal.

Release note: None

nvb committed:
roachtest: create TPC-C benchmarking tool

Closes #25084.

This change creates a new `tpccbench` roachtest specifically for
the purpose of benchmarking TPC-C. The test spins up a cluster (if
not provided a `--cluster` flag), restores a TPC-C dataset (if not
run on an existing cluster with `--wipe=false` specified), and
benchmarks the cluster's ability to scale under an increasing number
of TPC-C warehouses. Specifically, it uses a line search algorithm
to efficiently search for the maximum number of warehouses that can
be operated on while sustaining a throughput threshold, which is set
to a fraction of max tpmC for that warehouse count.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25310)
<!-- Reviewable:end -->

nvanbenschoten commented:
For reference, here's an example of the test logs when I ran the tool after bumping up the success threshold from 80% of max tpmC to 90% of max tpmC. This change results in my initial estimate being fairly inacurate, which is why it took so many iterations to converge:

<details><summary>test run logs (with some lines removed for clarity)</summary>

```
➜  cockroach git:(nvanbenschoten/tpccTool) ✗ ./bin/roachtest run tpccbench/nodes=3 --cluster=nathan-tpcc --wipe=false
=== RUN   tpccbench/nodes=3

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=380 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[   1] tpccbench/nodes=3: running benchmark, warehouses=380 (27s)
[   2] tpccbench/nodes=3: running benchmark, warehouses=380 (1m27s)
[   3] tpccbench/nodes=3: running benchmark, warehouses=380 (2m27s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3902.1  79.9%   3037.2   2684.4   5368.7   6979.3  13421.8  27917.3
--- FAIL: tpcc 380 resulted in 3902.100000 tpmC (80.2% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=375 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[   5] tpccbench/nodes=3: running benchmark, warehouses=375 (48s)
[   6] tpccbench/nodes=3: running benchmark, warehouses=375 (1m48s)
[   7] tpccbench/nodes=3: running benchmark, warehouses=375 (2m48s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3653.1  75.8%   4961.1   2684.4  11811.2  18253.6  34359.7  73014.4
--- FAIL: tpcc 375 resulted in 3653.100000 tpmC (76.1% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=365 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[   8] tpccbench/nodes=3: running benchmark, warehouses=365 (9s)
[   9] tpccbench/nodes=3: running benchmark, warehouses=365 (1m9s)
[  10] tpccbench/nodes=3: running benchmark, warehouses=365 (2m9s)
[  11] tpccbench/nodes=3: running benchmark, warehouses=365 (3m9s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     4057.1  86.4%   1437.8   1342.2   2684.4   3221.2   4563.4   9126.8
--- FAIL: tpcc 365 resulted in 4057.100000 tpmC (86.8% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=345 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[  12] tpccbench/nodes=3: running benchmark, warehouses=345 (30s)
[  13] tpccbench/nodes=3: running benchmark, warehouses=345 (1m30s)
[  14] tpccbench/nodes=3: running benchmark, warehouses=345 (2m30s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3927.1  88.5%   1032.3    805.3   2080.4   2550.1   3623.9   6174.0
--- FAIL: tpcc 345 resulted in 3927.100000 tpmC (88.9% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=305 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[  16] tpccbench/nodes=3: running benchmark, warehouses=305 (53s)
[  17] tpccbench/nodes=3: running benchmark, warehouses=305 (1m53s)
[  18] tpccbench/nodes=3: running benchmark, warehouses=305 (2m53s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3530.3  90.0%    316.4    218.1    738.2   1006.6   1476.4   3221.2
--- PASS: tpcc 305 resulted in 3530.300000 tpmC (90.4% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=325 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[  19] tpccbench/nodes=3: running benchmark, warehouses=325 (18s)
[  20] tpccbench/nodes=3: running benchmark, warehouses=325 (1m18s)
[  21] tpccbench/nodes=3: running benchmark, warehouses=325 (2m18s)
[  22] tpccbench/nodes=3: running benchmark, warehouses=325 (3m18s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3732.3  89.3%    643.1    436.2   1543.5   1811.9   2281.7   4831.8
--- FAIL: tpcc 325 resulted in 3732.300000 tpmC (89.7% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=315 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[  23] tpccbench/nodes=3: running benchmark, warehouses=315 (39s)
[  24] tpccbench/nodes=3: running benchmark, warehouses=315 (1m39s)
[  25] tpccbench/nodes=3: running benchmark, warehouses=315 (2m39s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3637.9  89.8%    511.7    385.9   1140.9   1409.3   1879.0   4026.5
--- PASS: tpcc 315 resulted in 3637.900000 tpmC (90.2% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=320 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[  26] tpccbench/nodes=3: running benchmark, warehouses=320 (1s)
[  27] tpccbench/nodes=3: running benchmark, warehouses=320 (1m1s)
[  28] tpccbench/nodes=3: running benchmark, warehouses=320 (2m1s)
[  29] tpccbench/nodes=3: running benchmark, warehouses=320 (3m1s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3690.3  89.7%    662.0    419.4   1610.6   2013.3   2684.4   6174.0
--- PASS: tpcc 320 resulted in 3690.300000 tpmC (90.1% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=322 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[  30] tpccbench/nodes=3: running benchmark, warehouses=322 (24s)
[  31] tpccbench/nodes=3: running benchmark, warehouses=322 (1m24s)
[  32] tpccbench/nodes=3: running benchmark, warehouses=322 (2m24s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3756.3  90.7%    444.3    352.3    939.5   1140.9   1476.4   3489.7
--- PASS: tpcc 322 resulted in 3756.300000 tpmC (91.1% of max tpmC)

> roachprod run nathan-tpcc:4 -- ./workload run tpcc --warehouses=323 --ramp=30s --duration=3m --split --scatter {pgurl:1-3}
[  34] tpccbench/nodes=3: running benchmark, warehouses=323 (49s)
[  35] tpccbench/nodes=3: running benchmark, warehouses=323 (1m49s)
[  36] tpccbench/nodes=3: running benchmark, warehouses=323 (2m49s)
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  180.0s     3701.6  89.1%    605.0    419.4   1409.3   1811.9   2952.8   7516.2
--- FAIL: tpcc 323 resulted in 3701.600000 tpmC (89.5% of max tpmC)

--------
MAX WAREHOUSES = 322
--------

--- PASS: tpccbench/nodes=3 (2273.68s)
PASS
```
</details>

nvb committed:
color: add missing color codes

This change adds color codes from black, green, blue, magenta and white.

Release note: None

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---

*[pkg/cmd/roachtest/tpcc.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LBhZ9GP64tqyLHHlTCW:-LBhZ9GP64tqyLHHlTCX:b-bxwz07) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L82)):*
> ```Go
> 				"sudo", "umount", "/mnt/data1", ";",
> 				"sudo", "mount", "-o", "discard,defaults,nobarrier",
> 				"/dev/disk/by-id/google-local-ssd-0", "/mnt/data1")
> ```

@We're going to want to use this elsewhere, so let's pull this out into a helper function: `remountNoBarrier(c *cluster)`.

---

*[pkg/cmd/roachtest/tpcc.go, line 217 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LBh_Z2DB7fi1i8vCdu2:-LBh_Z2DB7fi1i8vCdu3:b-dghm7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L217)):*
> ```Go
> 	}
> 
> 	r.Add(testSpec{
> ```

Do we want to run this every night? Or is this for ease of experimentation? If the latter, perhaps we should have a different `roachtest` command to invoke these non-nightly tests. For example, `roachtest bench tpcc`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25310)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/cmd/roachtest/tpcc.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LBhZ9GP64tqyLHHlTCW:-LBhawTd0n-ZB1WMNvKp:b-293q1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L82)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

@We're going to want to use this elsewhere, so let's pull this out into a helper function: `remountNoBarrier(c *cluster)`.
</blockquote></details>

Good idea. I'll do that.

---

*[pkg/cmd/roachtest/tpcc.go, line 217 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LBh_Z2DB7fi1i8vCdu2:-LBhaQBL3bQy3_8SSCzd:b2b7mj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L217)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Do we want to run this every night? Or is this for ease of experimentation? If the latter, perhaps we should have a different `roachtest` command to invoke these non-nightly tests. For example, `roachtest bench tpcc`.
</blockquote></details>

Yeah, I went back on forth on this, so I'm glad you pointed it out. I do think splitting benchmark "tests" into a separate namespace is a good idea and that's probably what I'll end up doing here. That said, I like the idea of running this every night so that it can't rot. I'm not sure whether that concern is worth the cost of running this nightly though.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25310)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/tpcc.go, line 80 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnpNAM2l9PAiQ7uGvw:-LCnpNAN9L-8MOWJgvk2:blhhs2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L80)):*
> ```Go
> 		if !c.isLocal() {
> 			c.Run(ctx, c.All(),
> 				"sudo", "umount", "/mnt/data1", ";",
> ```

Should we avoid failing the entire test if the cluster isn't using local SSD? Or is there value to enforcing that the cluster is using local SSD?

---

*[pkg/cmd/roachtest/tpcc.go, line 96 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnuGN8-DhEDxYh_xQd:-LCnuGN8-DhEDxYh_xQe:bd25m3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L96)):*
> ```Go
> 		m, ctx = errgroup.WithContext(ctx)
> 		m.Go(func() error {
> 			const maxWarehouses = 1000
> ```

I assume the plan is to have a few fixed fixture sizes (and initial estimates, step sizes, etc) to choose from at the command line?

---

*[pkg/cmd/roachtest/tpcc.go, line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnwKgX3Lf3hNOyCwD0:-LCnwKgX3Lf3hNOyCwD1:b-sne5de) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L187)):*
> ```Go
> 
> 				// Determine whether this means the test passed or not.
> 				const passRatio = 0.85
> ```

Why was 0.85? Isn't that kind of a low bar?

---

*[pkg/util/search/search.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnjTeO3i74U3rZZDkB:-LCnjTeO3i74U3rZZDkC:b-q1f4m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0a66db0ab12c6ded6746ef88d7c889b4f6b0b9/pkg/util/search/search.go#L23)):*
> ```Go
> 	// largest value that passes the provided predicate function. It is only
> 	// valid to call Search once for a given Searcher instance.
> 	Search(func(int) (bool, error)) (int, error)
> ```

`func(int) (bool, error)` seems worth being a named type

---

*[pkg/util/search/search.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnfuHs9Kg3vkag6JG7:-LCnfuHs9Kg3vkag6JG8:bjojh58) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0a66db0ab12c6ded6746ef88d7c889b4f6b0b9/pkg/util/search/search.go#L25)):*
> ```Go
> 	Search(func(int) (bool, error)) (int, error)
> 
> 	// The following two methods are un-exported and are by searchWithSearcher
> ```

s/are by/are used by/

---

*[pkg/util/search/search.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnjlVU-7DsptFS0fgf:-LCnjlVU-7DsptFS0fgg:b6z7bc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0a66db0ab12c6ded6746ef88d7c889b4f6b0b9/pkg/util/search/search.go#L60)):*
> ```Go
> 	if pass {
> 		if cur >= ss.max {
> 			panic("passed at index above max")
> ```

I'd recommend printing out `cur` and `ss` in these panic cases. And more generally including the failing values for all the panics in this file

---

*[pkg/util/search/search.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnlaGe5YVXe6fkHocF:-LCnlaGe5YVXe6fkHocG:b-xku08g) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0a66db0ab12c6ded6746ef88d7c889b4f6b0b9/pkg/util/search/search.go#L118)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 	if pass {
> 		bs.cur = mid(bs.cur, bs.ss.max)
> 	} else {
> 		bs.cur = mid(bs.ss.min, bs.cur)
> 	}
> ```
</details>


Can't this just be `bs.cur = mid(bs.ss.min, bs.ss.max)`? We've already updated the searchSpace bounds appropriately.

---

*[pkg/util/search/search.go, line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnwrkY99NawGoOMYVR:-LCnwrkY99NawGoOMYVS:blt3ysj) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/util/search/search.go#L81)):*
> ```Go
> }
> 
> // NewBinarySearcher returns a Searcher implementing a binary search strategy.
> ```

Why implement a binary searcher if we aren't going to use it? And if we really wanted binary search behavior, it seems as though we could just use `sort.Search`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25310)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I would also like to see this be run nightly to prevent rot, but a short
duration will be sufficient.--

-- Tobias

nvb committed:
roachtest: give Chaos a Stopper channel

This change replaces the Chaos duration with a channel that can be
hooked up to `time.After` to achieve the previous behavior. This allows
the chaos agent to be more flexible.

Release note: None

nvb committed:
roachtest: don't destroy constructed cluster when wipe=false

When the `--wipe` flag is set to false, it doesn't make sense to destroy a
cluster just because it was created by roachtest. This change makes the
`--wipe` flag instruct roachtest to leave a cluster alone after the
test, regardless of the `--cluster` flag. This is useful for allowing
roachtest to create a cluster that may then be used again for future
testing/benchmarking.

Release note: None

nvanbenschoten commented:
Review status: 1 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/cmd/roachtest/tpcc.go, line 80 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnpNAM2l9PAiQ7uGvw:-LD7pm1C06XS-MRNKk1Y:bsxoa8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L80)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Should we avoid failing the entire test if the cluster isn't using local SSD? Or is there value to enforcing that the cluster is using local SSD?
</blockquote></details>

For these benchmarks, I think there's a lot of value in clusters being set up exactly the same for each run, so for now I'd like to leave this as prescriptive as possible.

---

*[pkg/cmd/roachtest/tpcc.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LBhZ9GP64tqyLHHlTCW:-LD7q0DIAK-2cegMXhCC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L82)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Good idea. I'll do that.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/tpcc.go, line 96 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnuGN8-DhEDxYh_xQd:-LD7q17I3uy-fDSJnl07:b-taqnoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L96)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I assume the plan is to have a few fixed fixture sizes (and initial estimates, step sizes, etc) to choose from at the command line?
</blockquote></details>

Yep! Done.

---

*[pkg/cmd/roachtest/tpcc.go, line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnwKgX3Lf3hNOyCwD0:-LD7q6JS3uBrRHaAOEjU:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/cmd/roachtest/tpcc.go#L187)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Why was 0.85? Isn't that kind of a low bar?
</blockquote></details>

Added a comment.

---

*[pkg/util/search/search.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnjTeO3i74U3rZZDkB:-LD7uCt52SJMNp-em1m1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0a66db0ab12c6ded6746ef88d7c889b4f6b0b9/pkg/util/search/search.go#L23)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

`func(int) (bool, error)` seems worth being a named type
</blockquote></details>

Done.

---

*[pkg/util/search/search.go, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnfuHs9Kg3vkag6JG7:-LD7ugoK8F6X2MXZtzSZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0a66db0ab12c6ded6746ef88d7c889b4f6b0b9/pkg/util/search/search.go#L25)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

s/are by/are used by/
</blockquote></details>

Done.

---

*[pkg/util/search/search.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnjlVU-7DsptFS0fgf:-LD7uhEJ4nRqyhCH_x53:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0a66db0ab12c6ded6746ef88d7c889b4f6b0b9/pkg/util/search/search.go#L60)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I'd recommend printing out `cur` and `ss` in these panic cases. And more generally including the failing values for all the panics in this file
</blockquote></details>

Done.

---

*[pkg/util/search/search.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnlaGe5YVXe6fkHocF:-LD7w7Y_4DVjaZvzMd9K:b-taqnoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0a66db0ab12c6ded6746ef88d7c889b4f6b0b9/pkg/util/search/search.go#L118)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> ```
> 	if pass {
> 		bs.cur = mid(bs.cur, bs.ss.max)
> 	} else {
> 		bs.cur = mid(bs.ss.min, bs.cur)
> 	}
> ```

Can't this just be `bs.cur = mid(bs.ss.min, bs.ss.max)`? We've already updated the searchSpace bounds appropriately.
</blockquote></details>

Yep! Done.

---

*[pkg/util/search/search.go, line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LCnwrkY99NawGoOMYVR:-LD7uj2wDpkWdfzikaNo:b-yauyuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc08410f3501bf30afb26732cf2a83d03830737/pkg/util/search/search.go#L81)):*
> Why implement a binary searcher if we aren't going to use it?

It was helpful as a reference for the more complex `lineSearcher` implementation. It's also used to test `lineSearcher` in `TestIdenticalSearchers`, which gave me more confidence that the `lineSearcher` was correct.

I also originally had plans to use the binary searcher in tpccbench if no estimate was provided. It's not being used now, but it might be in the future.

> it seems as though we could just use sort.Search

`sort.Search` isn't easily usable for a few reasons:
1. it doesn't have a notion of precision, which is important for benchmarking
2. it return the smallest index i in [0, n) at which f(i) is true. Even if we invert the predicate, we'll have an off-by-one error on at the boundary, because we want the largest value that passes, not the smallest value that doesnt pass.
3. it doesn't fit into the `searchWithSearcher` model here that allows for code to be shared between the sorting algorithms.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25310)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 12 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/roachtest/cluster.go, line 753 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LD8XFmQE3y_XjdpeJOB:-LD8XFmQE3y_XjdpeJOC:bcyb9hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3890f2dfc709c85c467e57fc85e9fbb2f240aa21/pkg/cmd/roachtest/cluster.go#L753)):*
> ```Go
> 
> // remountNoBarrier remounts the cluster's local SSDs with the nobarrier option.
> func remountNoBarrier(ctx context.Context, c *cluster) {
> ```

Perhaps this should be `(*cluster).RemountNoBarrier`.

---

*[pkg/cmd/roachtest/cluster.go, line 754 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LD8XNu3E_hNw6BspE14:-LD8XNu3E_hNw6BspE15:b-1grn0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/3890f2dfc709c85c467e57fc85e9fbb2f240aa21/pkg/cmd/roachtest/cluster.go#L754)):*
> ```Go
> // remountNoBarrier remounts the cluster's local SSDs with the nobarrier option.
> func remountNoBarrier(ctx context.Context, c *cluster) {
> 	c.Stop(ctx, c.All())
> ```

This is a surprising side-effect. Probably better to have the test fail if the author forgot to stop the nodes before performing this operation.

---

*[pkg/cmd/roachtest/test.go, line 69 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LD8Xjde9AoTtSWHY8TB:-LD8Xjde9AoTtSWHY8TC:bxylvvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3890f2dfc709c85c467e57fc85e9fbb2f240aa21/pkg/cmd/roachtest/test.go#L69)):*
> ```Go
> 	// Benchmark indicates that the test is a benchmark and should not be run
> 	// unless the --bench flag is passed.
> 	Benchmark bool
> ```

Rather than a `--bench` flag, my instinct is that we should add a `bench` command and a `registerBenchmarks` function. Do you have any strong feelings about the approach you've taken here, or have arguments for why it is better? The ergonomics seem similar: `roachtest run --bench` vs `roachtest bench`. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25310)*
<!-- Sent from Reviewable.io -->

nvb committed:
roachtest: pull RemountNoBarrier into helper method

Release note: None

nvanbenschoten commented:
Review status: 1 of 11 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/cmd/roachtest/cluster.go, line 753 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LD8XFmQE3y_XjdpeJOB:-LD8rV7rEAsBwnqE0F61:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3890f2dfc709c85c467e57fc85e9fbb2f240aa21/pkg/cmd/roachtest/cluster.go#L753)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Perhaps this should be `(*cluster).RemountNoBarrier`.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/cluster.go, line 754 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LD8XNu3E_hNw6BspE14:-LD8rUlN2w3dQjs7xQpF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3890f2dfc709c85c467e57fc85e9fbb2f240aa21/pkg/cmd/roachtest/cluster.go#L754)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is a surprising side-effect. Probably better to have the test fail if the author forgot to stop the nodes before performing this operation.
</blockquote></details>

Done.

---

*[pkg/cmd/roachtest/test.go, line 69 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-LD8Xjde9AoTtSWHY8TB:-LD9MfUHB32ieCAt9Hm5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3890f2dfc709c85c467e57fc85e9fbb2f240aa21/pkg/cmd/roachtest/test.go#L69)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Rather than a `--bench` flag, my instinct is that we should add a `bench` command and a `registerBenchmarks` function. Do you have any strong feelings about the approach you've taken here, or have arguments for why it is better? The ergonomics seem similar: `roachtest run --bench` vs `roachtest bench`. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25310)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (for the roachtest infrastructure changes)

---

Review status: 1 of 11 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25310#-:-LD9bPZM9ImwHOtHjYHe:b4n0pbz)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
TFTRs!

bors r+

nvanbenschoten commented:
Failure is related to https://github.com/cockroachdb/cockroach/issues/25797.

bors r+
